library/std/src/sys/sync/rwlock/queue.rs

```diff
@@ -449,6 +462,88 @@ impl RwLock {
+            Release,
+            Relaxed,
+        ) {
+            Ok(_) => return,
+            Err(state) => debug_assert!(
+                state.mask(LOCKED).addr() != 0 && state.mask(QUEUED).addr() != 0,
+                "RwLock should be LOCKED and QUEUED"
+            ),
+        }
+
+        // Attempt to grab the queue lock.
+        let state = loop {
+            match self.state.fetch_update(Release, Acquire, |ptr: State| {
+                // Go from not queue locked to being queue locked.
+                if ptr.mask(QUEUE_LOCKED).addr() == 0 {
+                    Some(without_provenance_mut(ptr.addr() | QUEUE_LOCKED))
+                } else {
+                    None
+                }
+            }) {
+                Ok(state) => break state,
+                Err(_) => {}
+            }
+        };
+
+        // SAFETY: By Invariant 2 we know that this tail is valid.
+        let mut tail_ptr = unsafe { add_backlinks_and_find_tail(to_node(state)) };
+        let mut prev_ptr;
+
+        // We start with 1 reader, which is the current thread.
+        let mut readers = 1;
+
+        // Wake up all readers at the top of the queue.
+        loop {
+            // SAFETY: We have the queue lock so nobody else can be modifying the queue or the tail.
+            let tail = unsafe { tail_ptr.as_mut() };
+
+            // If the current `tail` is a writer thread, then we have no readers left to wake.
+            // If there is no valid `prev` backlink, then we don't want to remove this node either
+            // because we will have no `tail` that we can replace this node with.
+            if tail.write || tail.prev.get().is_none() {
+                // Store the reader count in the `next` field.
+                tail.next.0 = AtomicPtr::new(without_provenance_mut(readers));
+                break;
+            }
+
+            prev_ptr = tail.prev.get().expect("Cannot be a `None` variant by the above");
+
+            // SAFETY: By Invariant 2, we know that all `prev` backlinks are valid, and since we
+            // have the queue lock, nobody can be modifying the queue or this node.
+            let prev = unsafe { prev_ptr.as_mut() };
+
+            // Notify the reader thread that they are now in read mode.
+            tail.is_reading.store(true, Release);
+            readers += 1;
+            prev.next.0 = AtomicPtr::new(without_provenance_mut(readers));
+
+            // SAFETY: Because we observed that there were  waiting threads from the first CAS, we
+            // know that the state must have been a valid head to the wait queue as nobody can be
+            // modifying the node itself.
+            let head = unsafe { to_node(self.state.load(Acquire)).as_ref() };
+
+            // Split off the last node and update the `tail` field of `state`.
+            head.tail.set(Some(prev_ptr));
+            tail_ptr = prev_ptr;
+
+            unsafe {
+                Node::complete(tail_ptr);
+            }
```

Comment:
Right now, this `complete` call is somehow waking up writers, even though I check above (approx line 512) that the `tail` is not `write`. If anyone knows why this is happening, please let me know!